src/validation.cpp

```diff
@@ -2837,26 +2836,22 @@ bool Chainstate::FlushStateToDisk(
                 }
             }
         }
-        const auto nNow{SteadyClock::now()};
+        const auto nNow{NodeClock::now()};
         // Avoid writing/flushing immediately after startup.
         if (m_last_write == decltype(m_last_write){}) {
             m_last_write = nNow;
-        }
-        if (m_last_flush == decltype(m_last_flush){}) {
-            m_last_flush = nNow;
+            m_random_write_delay = std::chrono::minutes{(FastRandomContext().randrange(DATABASE_WRITE_INTERVAL_RANDOM_DELAY_RANGE))};
```

Comment:
given that this delay is written multiple times, would that be more verbose?
Alternatively maybe we could add an `m_next_write` next to `m_last_write`, like:
```patch
diff --git a/src/validation.cpp b/src/validation.cpp
index b54446be4e..e2c7272531 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -86,10 +86,9 @@ using node::CBlockIndexHeightOnlyComparator;
 using node::CBlockIndexWorkComparator;
 using node::SnapshotMetadata;
 
-/** Time to wait between writing blocks/block index and chainstate to disk. */
-static constexpr std::chrono::minutes DATABASE_WRITE_INTERVAL{50};
-/** Range of random delay to add to DATABASE_WRITE_INTERVAL */
-static constexpr uint32_t DATABASE_WRITE_INTERVAL_RANDOM_DELAY_RANGE{20};
+/** Minimum and maximum time to wait between writing blocks/block index and chainstate to disk. */
+static constexpr auto DATABASE_WRITE_INTERVAL_MIN{50min};
+static constexpr auto DATABASE_WRITE_INTERVAL_MAX{70min};
 /** Maximum age of our tip for us to be considered current for fee estimation */
 static constexpr std::chrono::hours MAX_FEE_ESTIMATION_TIP_AGE{3};
 const std::vector<std::string> CHECKLEVEL_DOC {
@@ -2776,6 +2775,11 @@ CoinsCacheSizeState Chainstate::GetCoinsCacheSizeState(
     return CoinsCacheSizeState::OK;
 }
 
+NodeClock::time_point Chainstate::CalculateNextWrite(NodeClock::time_point after) const
+{
+    return FastRandomContext().rand_uniform_delay(after + DATABASE_WRITE_INTERVAL_MIN, DATABASE_WRITE_INTERVAL_MAX - DATABASE_WRITE_INTERVAL_MIN);
+}
+
 bool Chainstate::FlushStateToDisk(
     BlockValidationState &state,
     FlushStateMode mode,
@@ -2840,14 +2844,14 @@ bool Chainstate::FlushStateToDisk(
         // Avoid writing/flushing immediately after startup.
         if (m_last_write == decltype(m_last_write){}) {
             m_last_write = nNow;
-            m_random_write_delay = std::chrono::minutes{(FastRandomContext().randrange(DATABASE_WRITE_INTERVAL_RANDOM_DELAY_RANGE))};
+            m_next_write = CalculateNextWrite(m_last_write);
         }
         // The cache is large and we're within 10% and 10 MiB of the limit, but we have time now (not in the middle of a block processing).
         bool fCacheLarge = mode == FlushStateMode::PERIODIC && cache_state >= CoinsCacheSizeState::LARGE;
         // The cache is over the limit, we have to write now.
         bool fCacheCritical = mode == FlushStateMode::IF_NEEDED && cache_state >= CoinsCacheSizeState::CRITICAL;
         // It's been a while since we wrote the block index to disk. Do this frequently, so we don't need to redownload after a crash.
-        bool fPeriodicWrite = mode == FlushStateMode::PERIODIC && nNow > m_last_write + DATABASE_WRITE_INTERVAL + m_random_write_delay;
+        bool fPeriodicWrite = mode == FlushStateMode::PERIODIC && nNow >= m_next_write;
         // Combine all conditions that result in a full cache flush.
         bool fDoFullFlush = (mode == FlushStateMode::ALWAYS) || fCacheLarge || fCacheCritical || fPeriodicWrite || fFlushForPrune;
         // Write blocks, block index and best chain related state to disk.
@@ -2882,7 +2886,7 @@ bool Chainstate::FlushStateToDisk(
                 m_blockman.UnlinkPrunedFiles(setFilesToPrune);
             }
             m_last_write = nNow;
-            m_random_write_delay = std::chrono::minutes{(FastRandomContext().randrange(DATABASE_WRITE_INTERVAL_RANDOM_DELAY_RANGE))};
+            m_next_write = CalculateNextWrite(m_last_write);
 
             if (!CoinsTip().GetBestBlock().IsNull()) {
                 LOG_TIME_MILLIS_WITH_CATEGORY(strprintf("write coins cache to disk (%d coins, %.2fkB)",
diff --git a/src/validation.h b/src/validation.h
index dfe33b8e29..953793bdf7 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -810,7 +810,9 @@ private:
         EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
     NodeClock::time_point m_last_write{};
-    std::chrono::minutes m_random_write_delay{};
+    NodeClock::time_point m_next_write{};
+
+    NodeClock::time_point CalculateNextWrite(NodeClock::time_point after) const;
 
     /**
      * In case of an invalid snapshot, rename the coins leveldb directory so
```

note: this way it should probably be `nNow >= m_next_write` instead of `nNow > m_next_write` to be consistent with the naming.

Reply:
Thanks for the suggestions! I could not get @sipa's to compile, so went with @l0rinc's.